### PR TITLE
Problem (WIP #1996): no support for genesis build procedure involving TDBE

### LINKED ISSUE #1996: Problem: no support for genesis build procedure involving TDBE

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- preliminary try implemented in build.sh

### CODE REVIEW

**Review by tomtau — COMMENTED** [review]
bors try

File: chain-tx-enclave-next/tdbe/tdbe-common/src/lib.rs [hunk 1]
```diff
@@ -10,3 +11,10 @@ pub struct TdbeConfig {
     /// Transaction IDs to fetch from another TDBE server
     pub transaction_ids: Vec<TxId>,
 }
+
+const LIGHT_GENESIS: &str = env!("LIGHT_GENESIS");
```
Comment (tomtau) [inline chain-tx-enclave-next/tdbe/tdbe-common/src/lib.rs:15]: if "light_genesis" is in json, it'll be fairly error-prone (escaping-characters, length limitations on environment variables, etc.) -- `include_str!` may be better

File: chain-tx-enclave-next/tx-validation-next/src/sgx_module.rs [hunk 2]
```diff
@@ -19,6 +19,7 @@ use std::sync::Arc;
 
 /// FIXME: genesis app hash etc.?
 pub const NETWORK_HEX_ID: u8 = get_network_id!();
+pub const TDBE_MRENCLAVE: &str = env! {"TDBE_MRENCLAVE"};
```
Comment (tomtau) [inline chain-tx-enclave-next/tx-validation-next/src/sgx_module.rs:22]: maybe similar to https://github.com/crypto-com/chain/issues/2174 ?
```suggestion
pub const TDBE_MRENCLAVE: [u8; 32] = include_bytes! {"tdbe.mrenclave"};
```

File: chain-tx-enclave-next/tdbe/tdbe-common/src/lib.rs (hunk 1 above)
Comment (yihuang) [reply chain-tx-enclave-next/tdbe/tdbe-common/src/lib.rs]: done

**Comment by tomtau** [comment]
bors try

**Review by tomtau — APPROVED** [review]

**Review by tomtau — COMMENTED** [review]
@yihuang for the genesis build procedure, there may still be a few things missing:
* generating the keypackage and sealing the secrets -- currently in the integration tests, this is done via "mls" enclave, but that should be removed -- instead `TdbeStartupConfig` can have option for this where it'll generate the keypackage + it'll also send back the sealed state with secrets
* for the group creator, there should also be`TdbeStartupConfig` option to generate genesis-time Welcome/Commit/Add payloads (and provide the other nodes' keypackages)
* other nodes TDBE should be able to process those handshake payloads
* when starting up, recovered from that sealed genesis state / check it

ok to merge this PR, but it doesn't fix #1996 yet

**Review by tomtau — COMMENTED** [review]
@yihuang for the genesis build procedure, there may still be a few things missing:

* generating the keypackage and sealing the secrets -- currently in the integration tests, this is done via "mls" enclave, but that should be removed -- instead TdbeStartupConfig can have option for this where it'll generate the keypackage + it'll also send back the sealed state with secrets
* for the group creator, there should also beTdbeStartupConfig option to generate genesis-time Welcome/Commit/Add payloads (and provide the other nodes' keypackages)
* other nodes TDBE should be able to process those handshake payloads
* when starting up, recovered from that sealed genesis state / check it

ok to merge this PR, but it doesn't fix #1996 yet

**Review by leejw51crypto — APPROVED** [review]
lgtm

File: chain-tx-enclave-next/tx-validation-next/src/sgx_module.rs [hunk 3]
```diff
@@ -19,6 +19,7 @@ use std::sync::Arc;
 
 /// FIXME: genesis app hash etc.?
 pub const NETWORK_HEX_ID: u8 = get_network_id!();
+pub const TDBE_MRENCLAVE: &[u8; 32] = include_bytes!("tdbe.mrenclave");
```
Comment (tomtau) [inline chain-tx-enclave-next/tx-validation-next/src/sgx_module.rs:22]: `get_tdbe_mrenclave` proc macro can be removed (in the enclave-macro)
and this call can be replaced with `TDBE_MRENCLAVE`
https://github.com/crypto-com/chain/blob/master/chain-tx-enclave-next/tx-validation-next/src/sgx_module.rs#L38

Comment (yihuang) [reply chain-tx-enclave-next/tx-validation-next/src/sgx_module.rs:27]: done

**Pull request merged**